### PR TITLE
Fix typing annotations for the ipu strategy

### LINKED ISSUE #13445: Improve typing coverage (4/n)
## 🚀 Typing coverage

Let's improve typing coverage of PyTorch Lightning together!

I'm creating a new issue in order to increase visibility. There are three older issues (#7037, #5023, #4698) which became stale over time.

### Plan

Currently, there are 55 files which are excluded from mypy checks so that our CI does not fail. These files vastly differ in difficulty in order to make the typing complete. For this reason, we are introducing difficulty estimate for each file so that community members can choose to work on the files appropriate to their skill level.

Please, comment on this issue in order to reserve a particular file to work on. Once you do so, I will edit this top comment to avoid collisions. Once you think your work is finished, please open a PR referencing this issue which:

 * removes the corresponding line from `pyproject.toml`
 * and passes `mypy` checks with the corresponding line removed. You can test it locally by running `mypy` from root directory

If you are struggling with pushing it over the finish line, open the PR anyway and someone from our team will help you to get it there. 🚀

Please note, that it can happen that you may need to edit more than just one file. This is fine, but please keep in mind, that the goal of your PR will be to make the check passing for the chosen file. Also, please note that the difficulty is just an educated guess.

For those of you who are not familiar with the process of contributing a PR, we have prepared a [simple guide](https://devblog.pytorchlightning.ai/quick-contribution-guide-86d977171b3a) that will walk you through the necessary steps. You can do it! :rocket: :muscle: 

#### Difficulty 3 of 3

- [x] `pytorch_lightning/callbacks/progress/rich_progress.py` @donlapark #14963 

### Completed

<details>

#### Difficulty 1 of 3

- [x] `pytorch_lightning/core/decorators.py` #14044
- [x] `pytorch_lightning/profilers/advanced.py` @nninept #13792 
~- [ ] `pytorch_lightning/profilers/base.py` @LeeChanHyuk #13879~
- [x] `pytorch_lightning/loggers/base.py` @JustinGoheen #13494 
- [x] `pytorch_lightning/__setup__.py` @CyprienRicque #13472 
~- [ ] `pytorch_lightning/distributed/dist.py` @puhuk #13492~
- [x] `pytorch_lightning/strategies/single_device.py` @CyprienRicque #13532 
- [x] `pytorch_lightning/trainer/optimizers.py` @gautierdag #13470
- [x] `pytorch_lightning/utilities/distributed.py` @krishnakalyan3 #13678
- [x] `pytorch_lightning/callbacks/finetuning.py` @ar90n #13516 
- [x] `pytorch_lightning/loggers/mlflow.py` @JustinGoheen ~~#13690~~ #13691
- [x] `pytorch_lightning/tuner/tuning.py` @donlapark ~~#13616~~ #13631 
- [x] `pytorch_lightning/strategies/single_tpu.py` @CyprienRicque #13534 
- [x] `pytorch_lightning/strategies/ddp2.py` @CyprienRicque #13535 
- [x] `pytorch_lightning/strategies/parallel.py` @CyprienRicque #13556 
- [x] `pytorch_lightning/loggers/csv_logs.py` @JustinGoheen #13538 
- [x] `pytorch_lightning/tuner/lr_finder.py` @donlapark #13513 #13652 
- [x] `pytorch_lightning/strategies/dp.py` @CyprienRicque #13564 
- [x] `pytorch_lightning/profilers/simple.py` @krishnakalyan3 #14103
- [x] `pytorch_lightning/strategies/sharded_spawn.py` @krishnakalyan3 #14102
- [x] `pytorch_lightning/demos/mnist_datamodule.py` @alro923 #13929 
- [x] `pytorch_lightning/demos/boring_classes.py` @krishnakalyan3 #14201
- [x] `pytorch_lightning/tuner/batch_size_scaling.py` @ar90n #13518 

#### Difficulty 2 of 3

- [x] `pytorch_lightning/loops/epoch/training_epoch_loop.py` @himkt #13555 
- [x] `pytorch_lightning/core/mixins/device_dtype_mixin.py` @krishnakalyan3 #13704 
- [x] `pytorch_lightning/loggers/comet.py` @JustinGoheen #13689 
- [x] `pytorch_lightning/loggers/tensorboard.py` @JustinGoheen #13688 
- [x] `pytorch_lightning/strategies/horovod.py` @CyprienRicque #13570 
- [x] `pytorch_lightning/callbacks/model_checkpoint.py` @BongYang #13617
- [x] `pytorch_lightning/strategies/fully_sharded.py` @BongYang #13941 
- [x] `pytorch_lightning/loggers/neptune.py` @JustinGoheen #13692 
- [x] `pytorch_lightning/utilities/meta.py` @nninept #13763 #13868
- [x] `pytorch_lightning/strategies/tpu_spawn.py` @BongYang #13813  
- [x] `pytorch_lightning/loggers/logger.py` @JustinGoheen #13541 
- [x] `pytorch_lightning/loggers/wandb.py` @gautierdag #13483
- [x] `pytorch_lightning/callbacks/stochastic_weight_avg.py` @donlapark #13685 #13860
- [x] `pytorch_lightning/strategies/strategy.py` @CyprienRicque #13519 
- [x] `pytorch_lightning/strategies/deepspeed.py` @donlapark #13832 
- [x] `pytorch_lightning/strategies/ddp_spawn.py` @donlapark #13865 
- [x] `pytorch_lightning/strategies/ipu.py` @HalestormAI #13786
- [x] `pytorch_lightning/trainer/connectors/callback_connector.py` @krishnakalyan3 #13750
- [x] `pytorch_lightning/strategies/ddp.py` @lijm1358 #13885 
- [x] `pytorch_lightning/core/saving.py` @JustinGoheen #13932 
- [x] `pytorch_lightning/callbacks/quantization.py` @krishnakalyan3 #13782 
- [x] `pytorch_lightning/strategies/sharded.py` @lijm1358 #14184 
- [x] `pytorch_lightning/core/datamodule.py` @JustinGoheen #13693

#### Difficulty 3 of 3
~- [ ] `pytorch_lightning/trainer/callback_hook.py` @JustinGoheen #13807 ~
- [x] `pytorch_lightning/core/module.py` @JustinGoheen #13603 
- [x] `pytorch_lightning/trainer/connectors/data_connector.py` @JustinGoheen #13806
- [x] `pytorch_lightning/utilities/auto_restart.py` @donlapark #13904 
- [x] `pytorch_lightning/trainer/supporters.py` @donlapark #14633 
- [x] `pytorch_lightning/profilers/pytorch.py` @krishnakalyan3 #14405 
- [x] `pytorch_lightning/utilities/data.py` @nandwalritik #13901 
- [x] `pytorch_lightning/trainer/trainer.py` ~@JustinGoheen #13810~ @BongYang #14204 

</details>

cc @borda @justusschock @awaelchli @rohitgr7 @Borda @tchaton @aniketmaurya @kingjuno @alat-rights @carmocca @akihironitta @stancld as you were all involved in previous issues

### PR DESCRIPTION
**Pull request by HalestormAI**
Fixes mypy errors attributed to `pytorch_lightning/strategies/ipu.py` for issue https://github.com/Lightning-AI/lightning/issues/13445

    removed module name from pyproject.toml [tool.mypy.overrides]
    resolved discrepencies for `Optional` returned objects
    fixed missing and incorrect type hints.
    
Part of #13445

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by HalestormAI** [comment]
Hmm, ok the mypy test was passing locally, perhaps missed something when I rebased. Will fix.

**Comment by otaj** [comment]
Hi @HalestormAI, do you need some help with finalizing this PR? :muscle:

**Comment by HalestormAI** [comment]
Yes please - went round in circles a few times trying to resolve the typing clash on `lightning_module` - haven't yet found a way to do it. Been tied up with work things last week so haven't had much chance to revisit

File: src/pytorch_lightning/strategies/ipu.py [hunk 1]
```diff
@@ -320,24 +337,24 @@ def teardown(self) -> None:
...
 
-    def _detach_models(self):
+    def _detach_models(self) -> None:
         """Detaches all stage specific models from IPU devices."""
         for k, model in self.poptorch_models.items():
             if self._compiled(model) and model.isAttachedToDevice():
                 model.detachFromDevice()
 
-    def _load_model(self, stage: str):
+    def _load_model(self, stage: str) -> None:
```
Comment (otaj) [inline src/pytorch_lightning/strategies/ipu.py:357]: The stage could be typed as a `RunningStage` - and then `self.poptorch_models` can have keys as only `RunningStage` and not a `Union[RunningStage, str]`

File: src/pytorch_lightning/strategies/ipu.py [hunk 2]
```diff
@@ -112,12 +117,12 @@ def __init__(
         self.device_iterations = device_iterations
         self.autoreport = autoreport
         self.autoreport_dir = autoreport_dir
-        self.poptorch_models = {}
+        self.poptorch_models: Dict[Union[RunningStage, str], "poptorch.PoplarExecutor"] = {}
```
Comment (otaj) [inline src/pytorch_lightning/strategies/ipu.py:120]: Let's remove that `str` from the dict keys

**Review by otaj — APPROVED** [review]
LGTM! I added a small commit resolving those last couple errors. We'll see how the IPU CI likes it :muscle:

**Comment by otaj** [comment]
Alright, so my quick'n'dirty fix did not seem to work as it broke two IPU tests (which do make sense so we don't to get rid of them or rewrite them). So, seems like `lightning_module` has to stay optional.

**Comment by otaj** [comment]
Adding a bunch of asserts at correct places seemed to help, so now we're green with the tests again :tada:

**Comment by HalestormAI** [comment]
Great! Thanks @otaj . I was trying to avoid all the asserts, but I guess that's the only way!

**Comment by HalestormAI** [comment]
Should I update branch then create review? Do you prefer merge or rebase to line this up with master?

**Comment by otaj** [comment]
We are "merge" folks :) There's also a button "Update branch" (I don't know if you have it as well), which will merge master for you if there are no conflicts :)

**Comment by HalestormAI** [comment]
Thanks @otaj - I've updated the branch and fixed the suggestion above. Once the CI completes I'll put it in for review

File: src/pytorch_lightning/strategies/ipu.py [hunk 3]
```diff
@@ -220,15 +231,20 @@ def inference_opts(self) -> "poptorch.Options":
...
-        self, dataloader: DataLoader, sampler, mode: Optional[RunningStage] = None
+        self, dataloader: DataLoader, sampler: Union[Sampler, Iterable], mode: Optional[RunningStage] = None
     ) -> "poptorch.DataLoader":
         if isinstance(dataloader, poptorch.DataLoader):
             # the user is returning the `poptorch.DataLoader` directly, don't change anything.
             return dataloader
 
         dl_args, dl_kwargs = _get_dataloader_init_args_and_kwargs(
-            dataloader, sampler, mode, self.replication_factor > 1
+            dataloader, sampler, mode, self.replication_factor > 1  # type:ignore [arg-type]
```
Comment (awaelchli) [inline src/pytorch_lightning/strategies/ipu.py:247]: Let's list the args vertically, so it becomes clear which argument we are ignoring:
```suggestion
            dataloader, 
            sampler,  # type: ignore[arg-type]
            mode, 
            self.replication_factor > 1
```

**Review by awaelchli — APPROVED** [review]
Solid

File: src/pytorch_lightning/strategies/ipu.py [hunk 4]
```diff
@@ -139,6 +144,9 @@ def setup(self, trainer: "pl.Trainer") -> None:
 
         super().setup(trainer)
 
+        assert self.lightning_module is not None
+        assert self.lightning_module.trainer is not None
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/ipu.py:148]: This is not required

```suggestion
```

File: src/pytorch_lightning/strategies/ipu.py [hunk 5]
```diff
@@ -192,12 +200,15 @@ def replication_factor(self) -> int:
             if self._inference_opts:
                 return self._inference_opts.replication_factor
 
-            return len(self.parallel_devices)
-
+            return len(self.parallel_devices) if self.parallel_devices else 0
+        assert self.lightning_module.trainer is not None
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/ipu.py:204]: This is not required

```suggestion
```

File: src/pytorch_lightning/strategies/ipu.py [hunk 6]
```diff
@@ -192,12 +200,15 @@ def replication_factor(self) -> int:
...
         stage = self.lightning_module.trainer.state.stage
+        assert stage is not None
         return self.poptorch_models[stage]._options.toDict()["replication_factor"]
 
     def _create_opts(self, training: bool) -> "poptorch.Options":
+        assert self.lightning_module is not None
+        assert self.lightning_module.trainer is not None
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/ipu.py:211]: This is not required

```suggestion
```

File: src/pytorch_lightning/strategies/ipu.py [hunk 7]
```diff
@@ -240,6 +256,8 @@ def _handle_gradient_accumulation_steps(self) -> None:
 
         ``optimizer_step`` will be called on every batch, and the IPU will handle grad accumulation internally.
         """
+        assert self.lightning_module is not None
+        assert self.lightning_module.trainer is not None
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/ipu.py:260]: This is not required

```suggestion
```

File: src/pytorch_lightning/strategies/ipu.py [hunk 8]
```diff
@@ -220,15 +231,20 @@ def inference_opts(self) -> "poptorch.Options":
             self._inference_opts = self._create_opts(training=False)
         return self._inference_opts
 
+    @property
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/ipu.py:234]: Why is this added?

File: src/pytorch_lightning/strategies/ipu.py [hunk 9]
```diff
@@ -192,12 +195,13 @@ def replication_factor(self) -> int:
             if self._inference_opts:
                 return self._inference_opts.replication_factor
 
-            return len(self.parallel_devices)
-
+            return len(self.parallel_devices) if self.parallel_devices else 0
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/ipu.py:198]:
```suggestion
            assert self.parallel_devices
            return len(self.parallel_devices)
```

File: src/pytorch_lightning/strategies/ipu.py [hunk 10]
```diff
@@ -318,26 +326,26 @@ def teardown(self) -> None:
             # undo dataloader patching
             pl.trainer.connectors.data_connector._update_dataloader = self._update_dataloader_original
 
-        if self._optimizer_zero_grad_original is not None:
+        if self._optimizer_zero_grad_original is not None and self.lightning_module is not None:
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/ipu.py:329]:
```suggestion
        assert self.lightning_module is not None
        if self._optimizer_zero_grad_original is not None:
```

**Review by rohitgr7 — APPROVED** [review]
